ofrak_core/ofrak/core/entropy/entropy.c

```diff
@@ -92,8 +103,61 @@ int entropy(uint8_t *data, size_t data_len, uint8_t *result, size_t window_size,
             }
         }
 
-        log_percent((i * 100) / data_len);
+        log_percent((i * 100) / data_len, py_log_callback);
     }
 
     return 0;
 }
+
+
+PyObject* entropy_wrapper(PyObject* _, PyObject* args){
+    Py_buffer data_buffer;
+    size_t window_size;
+    PyObject* py_log_percent;
+
+    if (!PyArg_ParseTuple(args, "y*nO", &data_buffer, &window_size, &py_log_percent)){
+        return NULL;
+    }
+
+    uint8_t *data = data_buffer.buf;
+    size_t result_size = data_buffer.len - window_size;
+    uint8_t *result = (uint8_t*) calloc(result_size, sizeof(uint8_t));
+
+    // Actual entropy calculation
+    entropy(data, data_buffer.len, result, window_size, py_log_percent);
+
+    PyObject* result_object = Py_BuildValue("y#", result, result_size);
+
+    // Clean up memory
+    PyBuffer_Release(&data_buffer);
+    free(result);
```

Comment:
No, I read from the docs specifically that `Py_BuildValue` copies buffer-like objects, so in a case like this, you need to free the malloc'd buffer you passed to it (and are free to do so).

The relevant docs page:
https://docs.python.org/3.7/c-api/arg.html#c.Py_BuildValue

The memory safety here is important and good to double check. The Python developers seem to agree and try to make it very clear:

> In other words, if your code invokes malloc() and passes the allocated memory to Py_BuildValue(), your code is responsible for calling free() for that memory once Py_BuildValue() returns.

And about the `data_buffer`:

> However, when a Py_buffer structure gets filled, the underlying buffer is locked so that the caller can subsequently use the buffer even inside a Py_BEGIN_ALLOW_THREADS block without the risk of mutable data being resized or destroyed. As a result, you have to call PyBuffer_Release() after you have finished processing the data (or in any early abort case).